### ISSUE TITLE
nghttp2-1.9.2 Build Failure

### ISSUE THREAD

**Issue by eleshar** [nghttp2/nghttp2#573]
I'm attempting to build nghttp2-1.9.2 but keeps failing on Linux lsdev.biz 3.16.0-4-amd64 #1 SMP Debian 3.16.7-ckt25-1 (2016-03-06) x86_64
OpenSSL 1.0.2g  1 Mar 2016

```
make[3]: Entering directory '/root/sources/nghttp2-1.9.2/src'
  CXX      libnghttpx_a-util.o
  CXX      libnghttpx_a-http2.o
  CC       libnghttpx_a-timegm.o
  CXX      libnghttpx_a-app_helper.o
  CXX      libnghttpx_a-ssl.o
ssl.cc: In function ‘void nghttp2::ssl::{anonymous}::ssl_locking_cb(int, int, const char*, int)’:
ssl.cc:61:14: error: ‘CRYPTO_LOCK’ was not declared in this scope
   if (mode & CRYPTO_LOCK) {
              ^
ssl.cc: In function ‘void nghttp2::ssl::libssl_init()’:
ssl.cc:155:3: warning: ‘void OPENSSL_config(const char*)’ is deprecated (declared at /usr/local/include/openssl/conf.h:140) [-Wdeprecated-declarations]
   OPENSSL_config(nullptr);
   ^
ssl.cc:155:25: warning: ‘void OPENSSL_config(const char*)’ is deprecated (declared at /usr/local/include/openssl/conf.h:140) [-Wdeprecated-declarations]
   OPENSSL_config(nullptr);
                         ^
Makefile:1536: recipe for target 'libnghttpx_a-ssl.o' failed
make[3]: *** [libnghttpx_a-ssl.o] Error 1
make[3]: Leaving directory '/root/sources/nghttp2-1.9.2/src'
Makefile:2416: recipe for target 'all-recursive' failed
make[2]: *** [all-recursive] Error 1
make[2]: Leaving directory '/root/sources/nghttp2-1.9.2/src'
Makefile:566: recipe for target 'all-recursive' failed
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory '/root/sources/nghttp2-1.9.2'
Makefile:475: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by tatsuhiro-t** [comment]
Guessing from the log, you probably installed openssl by yourself.  CRYPTO_LOCK is defined in <openssl/crypto.h> header file.  Could you check it out?  Probably this kind of error comes from the intermixed usage of default OpenSSL which comes with Debian and locally installed one.

**Comment by eleshar** [comment]
@tatsuhiro-t Thank you for the reply, yes I installed OpenSSL myself.

I've checked config.h but not seeing anything for CRYPTO_LOCK

**Comment by tatsuhiro-t** [comment]
Looks like my previous comment has bad markup.  CRYPTO_LOCK is defined in `<openssl/crypto.h>`.

**Comment by eleshar** [comment]
@tatsuhiro-t I've removed all the manually installed openssl packages and reinstalled from distribution package, but now running into another error:

```
make[3]: Entering directory '/root/sources/nghttp2/src'
  CXX      libnghttpx_a-util.o
  CXX      libnghttpx_a-http2.o
  CC       libnghttpx_a-timegm.o
  CXX      libnghttpx_a-app_helper.o
  CXX      libnghttpx_a-ssl.o
In file included from ssl.cc:25:0:
ssl.h:32:25: fatal error: openssl/ssl.h: No such file or directory
compilation terminated.
Makefile:1541: recipe for target 'libnghttpx_a-ssl.o' failed
make[3]: *** [libnghttpx_a-ssl.o] Error 1
make[3]: Leaving directory '/root/sources/nghttp2/src'
Makefile:2435: recipe for target 'all-recursive' failed
make[2]: *** [all-recursive] Error 1
make[2]: Leaving directory '/root/sources/nghttp2/src'
Makefile:567: recipe for target 'all-recursive' failed
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory '/root/sources/nghttp2'
Makefile:476: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by tatsuhiro-t** [comment]
I think you need to run ./configure again since the installation of openssl has been changed.

**Comment by ddennerline** [comment]
I am experiencing the same issue using OpenSSL 1.1.0-pre6-dev. I cannot downgrade because I need to disable some CipherSuites for test purposes. The latest version of OpenSSL uses a different set of locking APIs. For now, I have just commented out the lock callback code in src/ssl.cc to get past this error. I am not going to use any multi-threaded nghttp tool. Once you get past this error, two more fixes need to be applied (working on that now).

error: SSL_CTX* shrpx::ssl::create_ssl_context(const char*, const char*))
warning: void OPENSSL_config(const char*) is deprecated

It appears that ssl.cc will need to adjust its behavior based on header version. 

> OpenSSL now uses a new threading API. It is no longer necessary to set locking callbacks to use OpenSSL in a multi-threaded environment. There are two supported threading models: pthreads and windows threads. It is also possible to configure OpenSSL at compile time for "no-threads". The old threading API should no longer be used. The functions have been replaced with "no-op" compatibility macros.

**Comment by tatsuhiro-t** [comment]
Thank you for telling us.
I figured out the following fixes are required:

* remove CRYPTO_LOCK stuff (they are sorted out by openssl, and no application intervention is required, just like boringSSL)
* don't use OPENSSL_config
* use provided API to access BIO member

**Comment by tatsuhiro-t** [comment]
neverbleed also failed to compile but it should be resolved in their repository.

**Comment by tatsuhiro-t** [comment]
Fix committed via 6d22898
